Azure/sonic-buildimage
# [supervisord] Monitoring the critical processes with supervisord.

**- Why I did it**
Initially, we used Monit to monitor critical processes in each container. If one of critical processes was not running
or crashed due to some reasons, then Monit will write an alerting message into syslog periodically. If we add a new process
in a container, the corresponding Monti configuration file will also need to update. It is a little hard for maintenance. 

Currently we employed event listener of Supervisod to do this monitoring. Since processes in each container are managed by
Supervisord, we can only focus on the logic of monitoring. 

**- How I did it**
We borrowed the event listener of Supervisord to monitor critical processes in containers. The event listener will take
following steps if it was notified one of critical processes exited unexpectedly:

1. The event listener will first check whether the auto-restart mechanism was enabled for this container or not.  If auto-restart mechanism was enabled, event listener will kill the Supervisord process, which should cause the container to exit and subsequently get restarted.

2. If auto-restart mechanism was not enabled for this contianer, the event listener will enter a loop which will first sleep 1 minute and then check whether the process is running. If yes, the event listener exits. If no, an alerting message will be written into syslog.

**- How to verify it**
First, we need checked whether the auto-restart mechanism of a container was enabled or not by running the command `show feature status`. If enabled, one critical process should be selected and killed manually, then we need check whether the container will be restarted or not. 

Second, we can disable the auto-restart mechanism if it was enabled at step 1 by running the commnad `sudo config feature autorestart <container_name> disabled`. Then one critical process should be selected and killed. After that, we will see the alerting message which will appear in the syslog every 1 minute.

**- Which release branch to backport (provide reason below if selected)**



- [x ] 202006

**- Description for the changelog**



**- A picture of a cute animal (not mandatory but encouraged)**
